test/e2e/webhook_test.go

```diff
@@ -37,259 +40,640 @@ import (
+				g.Expect(hubClient.Get(ctx, types.NamespacedName{Name: crpName}, &crp)).Should(Succeed())
+				crp.Spec.Policy = &placementv1beta1.PlacementPolicy{
+					PlacementType:    placementv1beta1.PickNPlacementType,
+					NumberOfClusters: &numOfClusters,
+				}
+				err := hubClient.Update(ctx, &crp)
+				if k8sErrors.IsConflict(err) {
+					return err
+				}
+				var statusErr *k8sErrors.StatusError
+				g.Expect(errors.As(err, &statusErr)).To(BeTrue(), fmt.Sprintf("Update CRP call produced error %s. Error type wanted is %s.", reflect.TypeOf(err), reflect.TypeOf(&k8sErrors.StatusError{})))
+				Expect(statusErr.ErrStatus.Message).Should(MatchRegexp("placement type is immutable"))
+				return nil
+			}, testutils.PollTimeout, testutils.PollInterval).Should(Succeed())
+		})
 	})
 
-	It("should deny create CRP with invalid GVK", func() {
-		Eventually(func(g Gomega) error {
+	Context("mutation webhooks tests", func() {
+		BeforeAll(func() {
+			By("creating work resources")
+			createWorkResources()
+
 			// Create the CRP.
-			crp := placementv1beta1.ClusterResourcePlacement{
+			crp := &placementv1beta1.ClusterResourcePlacement{
 				ObjectMeta: metav1.ObjectMeta{
 					Name: crpName,
-					// Add a custom finalizer; this would allow us to better observe
-					// the behavior of the controllers.
-					Finalizers: []string{customDeletionBlockerFinalizer},
 				},
 				Spec: placementv1beta1.PlacementSpec{
-					ResourceSelectors: []placementv1beta1.ClusterResourceSelector{
-						{
-							Group:   "",
-							Kind:    "InvalidNamespace",
-							Version: "v1",
-							Name:    "invalid",
-						},
-					},
+					ResourceSelectors: workResourceSelector(),
 				},
 			}
 			By(fmt.Sprintf("creating placement %s", crpName))
-			err := hubClient.Create(ctx, &crp)
-			var statusErr *k8sErrors.StatusError
-			g.Expect(errors.As(err, &statusErr)).To(BeTrue(), fmt.Sprintf("Create CRP call produced error %s. Error type wanted is %s.", reflect.TypeOf(err), reflect.TypeOf(&k8sErrors.StatusError{})))
-			Expect(statusErr.ErrStatus.Message).Should(MatchRegexp(regexp.QuoteMeta("failed to get GVR of the selector")))
-			return nil
-		}, testutils.PollTimeout, testutils.PollInterval).Should(Succeed())
+			Expect(hubClient.Create(ctx, crp)).To(Succeed(), "Failed to create CRP %s", crpName)
+		})
+
+		AfterAll(func() {
+			// Ensure that the CRP and related resources are deleted after the tests.
+			ensureCRPAndRelatedResourcesDeleted(crpName, allMemberClusters)
+		})
+
+		It("should allow update CRP with nil policy and update fields with default values", func() {
+			Eventually(func(g Gomega) error {
+				var createdCRP placementv1beta1.ClusterResourcePlacement
+				g.Expect(hubClient.Get(ctx, types.NamespacedName{Name: crpName}, &createdCRP)).Should(Succeed(), "Failed to get CRP %s", crpName)
+				// Update the CRP with nil policy.
+				g.Expect(createdCRP.Spec.Policy).ToNot(BeNil(), "CRP should have a policy")
+				createdCRP.Spec.Policy = nil
```

Comment:
I don't think we should allow this, It means a PickN CRP can be converted to a PickAll CRP if user sets policy to nil. I'm hoping the validation webhook will catch it if that's the case